### ISSUE TITLE
SuperDirt not completing startup

### ISSUE THREAD

**Issue by yaxu** [musikinformatik/SuperDirt#69]
SuperDirt will often (maybe 50% of time) not finish initialising. It seems to read in all the samples but then not come up with the message saying it's listening to Tidal. I have to close scide, make sure scsynth and sclang processes are killed then reload, usually it's fine the second time. From talking ot others this appears to be a common experience, cross-platform.

**Comment by telephon** [comment]
can you test whether it will also fail in that way when you don't load any samples? It should also connect without loading them.

**Comment by yaxu** [comment]
Yes will do, for now here's a dump of where it freezes for me when it does load samples.

```
init_OSC
compiling class library...
initPassOne started
	NumPrimitives = 712
initPassOne done
	compiling dir: '/usr/local/share/SuperCollider/SCClassLibrary'
	compiling dir: '/usr/local/share/SuperCollider/Extensions'
	compiling dir: '/home/alex/.local/share/SuperCollider/downloaded-quarks/Dirt-Samples'
	compiling dir: '/home/alex/.local/share/SuperCollider/downloaded-quarks/Vowel'
	compiling dir: '/home/alex/.local/share/SuperCollider/downloaded-quarks/SuperDirt'
	pass 1 done
	numentries = 1186276 / 17283194 = 0.069
	5459 method selectors, 3166 classes
	method table size 18216560 bytes, big table size 138265552
	Number of Symbols 14036
	Byte Code Size 404167
	compiled 447 files in 0.36 seconds

Info: 3 methods are currently overwritten by extensions. To see which, execute:
MethodOverride.printAll

compile done
Class tree inited in 0.01 seconds
booting 57110
Cleaning up temp synthdefs...
Found 0 LADSPA plugins
Faust: supercollider.cpp: sc_api_version = 2
Faust: FaustJPverbRaw numControls=11
Faust: supercollider.cpp: sc_api_version = 2
Faust: FaustGreyholeRaw numControls=7
JackDriver: client name is 'SuperCollider'
SC_AudioDriver: sample rate = 44100.000000, driver's block size = 1024
JackDriver: connected  system:capture_1 to SuperCollider:in_1
JackDriver: connected  system:capture_2 to SuperCollider:in_2
JackDriver: connected  SuperCollider:out_1 to system:playback_1
JackDriver: connected  SuperCollider:out_2 to system:playback_2
SuperCollider 3 server ready.
JackDriver: max output latency 46.4 ms
WARNING: Extension in '/usr/local/share/SuperCollider/SCClassLibrary/Platform/linux/extMIDIOut.sc' overwrites Meta_MIDIClient:externalSources in main class library.
WARNING: Extension in '/usr/local/share/SuperCollider/SCClassLibrary/Platform/linux/extMIDIOut.sc' overwrites Meta_MIDIClient:externalDestinations in main class library.
WARNING: Extension in '/usr/local/share/SuperCollider/SCClassLibrary/Platform/linux/extMIDIOut.sc' overwrites Meta_MIDIIn:connectAll in main class library.
Intentional overwrites must be put in a 'SystemOverwrites' subfolder.


*** Welcome to SuperCollider 3.8.0. *** For help press Ctrl-D.
SCDoc: Indexing help-files...
Receiving notification messages from server localhost
Shared memory server interface initialized
loading synthdefs in /home/alex/.local/share/SuperCollider/downloaded-quarks/SuperDirt/classes/../synths/core-modules.scd
---- core synth defs loaded ----
loading synthdefs in /home/alex/.local/share/SuperCollider/downloaded-quarks/SuperDirt/classes/../synths/core-synths.scd
loading synthdefs in /home/alex/.local/share/SuperCollider/downloaded-quarks/SuperDirt/classes/../synths/default-synths.scd
loading synthdefs in /home/alex/.local/share/SuperCollider/downloaded-quarks/SuperDirt/classes/../synths/try-load-extra-synths.scd
loading synthdefs in /home/alex/.local/share/SuperCollider/downloaded-quarks/SuperDirt/classes/../synths/tutorial-synths.scd

loading 216 sample banks:
808 (6) 808bd (25) 808cy (25) 808hc (5) 808ht (5) 808lc (5) 808lt (5) 808mc (5) 808mt (5) 808oh (5) 808sd (25) 909 (1) ab (12) ade (10) ades2 (9) ades3 (7) ades4 (6) alex (2) alphabet (26) amencutup (32) armora (7) arp (2) arpy (11) auto (11) baa (7) baa2 (7) bass (4) bass0 (3) bass1 (30) bass2 (5) bass3 (11) bassdm (24) bassfoo (3) battles (2) bd (24) bend (4) bev (2) bin (2) birds (10) birds3 (19) bleep (13) blip (2) blue (2) bottle (13) breaks125 (2) breaks152 (1) breaks157 (1) breaks165 (1) breath (1) bubble (8) can (14) casio (3) cb (1) cc (6) chin (4) chink (1) circus (3) clak (2) click (4) clubkick (5) co (4) control (2) cosmicg (15) cp (2) cr (6) crow (4) d (4) db (13) diphone (38) diphone2 (12) dist (16) dork2 (4) dorkbot (2) dr (42) dr2 (6) dr55 (4) dr_few (8) drum (6) drumtraks (13) e (8) east (9) electro1 (13) erk (1) f (1) feel (7) feelfx (8) fest (1) fire (1) flick (17) fm (17) foo (27) future (17) gab (10) gabba (4) gabbaloud (4) gabbalouder (4) glasstap (3) glitch (8) glitch2 (8) gretsch (24) gtr (3) h (7) hand (17) hardcore (12) hardkick (6) haw (6) hc (6) hh (13) hh27 (13) hit (6) hmm (1) ho (6) hoover (6) house (8) ht (16) if (5) ifdrums (3) incoming (8) industrial (32) insect (3) invaders (18) jazz (8) jungbass (20) jungle (13) jvbass (13) kicklinn (1) koy (2) kurt (7) latibro (8) led (1) less (4) lighter (33) linnhats (6) lt (16) made (7) made2 (1) mash (2) mash2 (4) metal (10) miniyeah (4) moan (1) monsterb (6) moog (7) mouth (15) mp3 (4) msg (9) mt (16) mute (28) newnotes (15) noise (1) noise2 (8) notes (15) numbers (9) oc (4) odx (15) off (1) outdoor (6) pad (3) padlong (1) pebbles (1) perc (6) peri (15) pluck (17) popkick (10) print (11) proc (2) procshort (8) psr (30) rave (8) rave2 (4) ravemono (2) realclaps (4) reverbkick (1) rm (2) rs (1) sax (22) sd (2) seawolf (3) sequential (8) sf (18) sheffield (1) short (5) sid (12) sine (6) sitar (8) sn (52) space (18) speakspell (12) speech (7) speechless (10) speedupdown (9) stab (23) stomp (10) subroc3d (11) sugar (2) sundance (6) tabla (26) tabla2 (46) tablex (3) tacscan (22) tech (13) techno (7) tink (5) tok (4) toys (13) trump (11) ul (10) ulgab (5) uxay (3) v (6) voodoo (5) wind (10) wobble (1) world (3) xmas (1) yeah (31) 
... file reading complete. Required 435 MB of memory.


loading 71 sample banks:

replacing '909' (1)
909 (1) acid (3) airhorn (1) arc (1) atlast (3) bd8 (25) bd9 (10) bk (96) bks (8) blk (20) bls (13) cb8 (1) ch8 (1) cl8 (1) claus (16) cp8 (1) cp9 (4) cr9 (4) dhm (1) 
WARNING: File reading failed for path: '/home/alex/SparkleShare/spicule/extra-samples/dirty/note-077.wav'


dirty (88) donk (1) funky (8) 
replacing 'gab' (10)
gab (10) hc8 (5) hh9 (4) hits (16) ht8 (5) ht9 (8) jungbleep (1) jv (13) jvgabba (46) jvr (25) 
WARNING: File reading failed for path: '/home/alex/SparkleShare/spicule/extra-samples/jvrave/note-099.wav'



WARNING: File reading failed for path: '/home/alex/SparkleShare/spicule/extra-samples/jvrave/note-100.wav'


jvrave (64) jx (73) lc8 (5) liq (1) lt8 (5) lt9 (8) ma8 (1) mc8 (5) micron (17) mt8 (5) mt9 (8) narc (2) oh8 (5) oh9 (4) onoff (5) os (3) paper (1) pulse (61) rash (73) ravebass (13) rd9 (4) redflag (1) repetition (7) rs8 (1) rs9 (2) sb (3) sclap (25) scream (1) sd8 (25) sd9 (15) shackup (1) shakedown (1) sl (4) sometimes (1) sqrpad (3) sr (64) stabs (19) vocobass (13) xx (89) 
... file reading complete. Required 242 MB of memory.


loading 29 sample banks:
osa (8) osa1 (6) osa10 (2) osa11 (4) osa12 (4) osa13 (2) osa14 (4) osa2 (7) osa3 (6) osa4 (4) osa5 (6) osa6 (3) osa7 (3) osa8 (3) osa9 (3) osb (26) osc1 (4) osc2 (4) osc3 (3) osc4 (4) osc5 (3) osc6 (3) osc7 (4) osc8 (2) osc9 (4) osd1 (12) osd2 (19) osd3 (15) osd4 (12) 
... file reading complete. Required 445 MB of memory.

SCDoc: Indexed 1813 documents in 3.33 seconds
```

**Comment by telephon** [comment]
can you try an move the line `~dirt.start ...` before the load?

**Comment by efairbanks** [comment]
I am also experiencing this issue. It seemingly started happening randomly on my docker instances. Now I can't get them to succeed. I'm using the superdirt_startup.scd from this repo.

**Comment by telephon** [comment]
Is it connected to the error `WARNING: File reading failed for path:` or not at all?

**Comment by ericfairbanks** [comment]
@telephon The same thing happens to me, but I'm not causing that warning to show up. I seem to have temporarily "solved" the problem by only using SuperDirt synths. _(no samples)_ It seems that if I don't load samples, there are no issues.

It seems to me that the `s.sync;` called after `~dirt.loadSoundFiles;` is not working as intended. On the machine having issues, I can see that `~dirt.loadSoundFiles;` takes longer than it does on my machines where this is not an issue. Also, sometimes some of the samples are not loaded successfully. _(for example, `bd` and `sn` will make sound, but `sn:3` will not)_

Is it possible that sample loading and SuperDirt startup are interrupting each other?
  
 _(*EDIT:* After further research, it seems like `s.sync;` needs to be called _*immediately*_ following buffer allocation/operations, and that calling it after the `~dirt.loadSoundFiles;` method is going to do a whole lot of nothing. At least, this is how I understand it. It seems to me that some changes need to be made to SuperDirt so that it blocks when it loads files into buffers.)_

**Comment by yaxu** [comment]
@ericfairbanks Elsewhere I noticed you're using an older version of supercollider, I believe only 3.7+ are known to work well with superdirt.

I've reinstalled my laptop for 2018 but will try to replicate this issue on this fresh install tonight.

**Comment by efairbanks** [comment]
I'm fairly certain that this issue is not specific to 3.6.6, but I am aware that I should be on a higher version. I'm going to try to get SC building from source soon on this container rather than pulling it from aptitude, but it would be nice to have this fixed so that it works with the version of SC bundled with most Debian-based distros.

Alternatively, it would also be nice if Ubuntu/Debian would update their SuperCollider package every few years. ;]
  
Anyway, I'm currently toying around trying to find a fix for this. Will update with any progress.

**Comment by fdragovic** [comment]
haven't tested this, but i was having this issue and it seemed to go away
after cleaning up my samples... i noticed that superdirt is always happy
when i export signed 16-bit wavs but ignores some higher bitrate ones
completely (specifially with audacities "other uncompressed files" format,
and i think samplerate wasn't an issue)

**Comment by yaxu** [comment]
@efairbanks I think it's a different issue, 3.6.6 fails to load some samples, while 3.7+ loads all samples but sometimes fails to complete initialisation. Supercollider 3.7/3.8 are in recent debian-based distros (although not yet in linux mint, I've just discovered!).

**Comment by efairbanks** [comment]
@yaxu Gotcha. I'm rather perplexed by all this. Calling `server.sync;` after loading each individual sample does seem to solve the issue for me, but it slows down sample load unacceptably. However, after some investigation into how `server.sync;` works, it should be fine to call it after loading a bunch of samples.

I'm going to build latest SuperCollider and see if I can't reproduce the issue you're having. Didn't know 3.7/3.8 were in. The default docker Ubuntu image seems to be stuck with 3.6.6 for some reason.

**Comment by telephon** [comment]
The sync in the startup file (https://github.com/musikinformatik/SuperDirt/blob/master/superdirt_startup.scd#L20) was intended only as a precaution. You can just leave it out.

So if you comment out `s.sync`, do you still have this problem?

There may be an issue with sync that I'll try to reproduce. But for now, this might be a workaround.

**Comment by telephon** [comment]
P.S., generally speaking:

3.6 is really not really supported anymore (3.7 came out three years ago). 3.8 should work much better in general. And why not try 3.9? https://github.com/supercollider/supercollider/releases There have been a lot of fixes since then.

**Comment by yaxu** [comment]
I'm currently on 3.8.1, with 1.0-alpha installed via Quarks.gui
I've moved the call to `start` and removed the `sync` and startup still stalls often. I am loading over 1G of samples though.

```
(
// configure the sound server: here you could add hardware specific options
// see http://doc.sccode.org/Classes/ServerOptions.html
s.options.numBuffers = 1024 * 16; // increase this if you need to load more samples
s.options.memSize = 8192 * 16; // increase this if you get "alloc failed" messages
s.options.maxNodes = 1024 * 32; // increase this if you are getting drop outs and the message "too many nodes"
s.options.numOutputBusChannels = 2; // set this to your hardware output channel size, if necessary
s.options.numInputBusChannels = 2; // set this to your hardware input channel size, if necessary
// boot the server and start SuperDirt
s.waitForBoot {
	~dirt = SuperDirt(2, s); // two output channels, increase if you want to pan across more channels
	~dirt.start(57120, [0, 0, 0, 0, 0]);   // start listening on port 57120, create two orbits, each sending audio to channel 0. You can direct sounds to the orbits from tidal e.g. by: `# orbit "0 1 1"
	~dirt.loadSoundFiles;   // load samples (path containing a wildcard can be passed in)
	~dirt.loadSoundFiles("/home/alex/SparkleShare/spicule/extra-samples/*");
	~dirt.loadSoundFiles("/home/alex/SparkleShare/damu-samples/*");
	//s.sync; // wait for samples to be read
}
)
```

**Comment by yaxu** [comment]
Ignore my previous deleted message, I am tired+stupid (with the call to `start` moved the 'listening to tidal' message got a bit lost in the post window)

**Comment by telephon** [comment]
:) I never saw it - what was it?

**Comment by benyoung** [comment]
For what it's worth I have this issue even with SC 3.9, and I've basically always had it: dirt sometimes just doesn't listen for tidal.  Generally nothing that I can do fixes the problem except restarting the laptop, which almost always works. 

since "turn it off and back on again" is usually the first thing I try, and in this case it's effective, I haven't really tried too hard to do further troubleshooting. But I'm game, if anyone wants data.

**Comment by telephon** [comment]
@benyoung just to narrow down the problem: if you do not load any soundfiles (or use the latest 1.0-dev branch of superdirt), do you still get this problem sometimes?

And: you describe this as something that suddenly happens while you play. Is this true or does it happen only after you start tidal/superdirt?

**Comment by jarmitage** [comment]
Still experiencing this issue frequently FYI! Even after upgrading to SC 3.9.3

It does seem to be correlated with how much stress the system is under.

**Comment by telephon** [comment]
What superdirt version is it?

**Comment by jarmitage** [comment]
Now that the latest is working again I have updated - will let you know if the problem continues!

**Comment by telephon** [comment]
great! please reopen if you find it again.
